File: HamiltonianCycle.java
package backtracking;

import java.util.Arrays;

/**
 * ============================================================
 *  HAMILTONIAN CYCLE
 * ============================================================
 *  Given an undirected graph with N vertices, determine if it
 *  contains a Hamiltonian Cycle — a path that visits every
 *  vertex exactly once and returns to the starting vertex.
 *
 *  Graph Representation: Adjacency Matrix
 *    adj[i][j] = 1  →  edge exists between vertex i and j
 *    adj[i][j] = 0  →  no edge
 *
 *  Approach : Backtracking — build path vertex by vertex,
 *             backtrack when no valid next vertex exists
 *  Time     : O((N-1)!)  worst case  (NP-Complete problem)
 *  Space    : O(N)  — path array + recursion depth
 * ============================================================
 */
public class HamiltonianCycle {

    private static final int NOT_PLACED = -1;

    /**
     * Solves the Hamiltonian Cycle problem for the given graph.
     *
     * @param graph  N×N adjacency matrix
     * @return       true if a Hamiltonian Cycle exists
     */
    public static boolean solveHamiltonianCycle(int[][] graph) {
        int n = graph.length;
        int[] path = new int[n];
        Arrays.fill(path, NOT_PLACED);

        // Always start at vertex 0
        path[0] = 0;

        if (backtrack(graph, path, 1, n)) {
            printCycle(path);
            return true;
        }

        System.out.println("  No Hamiltonian Cycle exists in this graph.");
        return false;
    }

    /**
     * Core backtracking function.
     *
     * @param graph    Adjacency matrix
     * @param path     Current path of placed vertices
     * @param pos      Index of the next slot to fill in path[]
     * @param n        Total number of vertices
     * @return         true if a complete Hamiltonian Cycle is found
     */
    private static boolean backtrack(int[][] graph, int[] path, int pos, int n) {

        // ── BASE CASE ────────────────────────────────────────
        // All vertices placed — check if there's an edge back to start
        if (pos == n) {
            // The last vertex must connect back to vertex 0 (path[0])
            return graph[path[pos - 1]][path[0]] == 1;
        }

        // ── TRY EACH VERTEX AS THE NEXT IN THE PATH ──────────
        for (int vertex = 1; vertex < n; vertex++) {   // 0 is always the start

            // ── PHASE 1: CHOOSE ─────────────────────────────
            if (isSafe(graph, path, vertex, pos)) {

                path[pos] = vertex;   // place this vertex at position pos

                // ── PHASE 2: EXPLORE ────────────────────────
                if (backtrack(graph, path, pos + 1, n)) {
                    return true;  // found a complete cycle!
                }

                // ── PHASE 3: BACKTRACK ──────────────────────
                path[pos] = NOT_PLACED;   // un-place, try next vertex
            }
        }

        // No vertex worked in this position → backtrack to previous call
        return false;
    }

    /**
     * Returns true if 'vertex' can be placed at path[pos].
     *
     * Conditions:
     *   1. There must be an edge from the previously placed vertex to 'vertex'
     *   2. 'vertex' must not already be in the path
     */
    private static boolean isSafe(int[][] graph, int[] path, int vertex, int pos) {
        // Condition 1: edge from path[pos-1] → vertex must exist
        if (graph[path[pos - 1]][vertex] == 0) {
            return false;
        }
        // Condition 2: vertex must not already be in path[0..pos-1]
        for (int i = 0; i < pos; i++) {
            if (path[i] == vertex) {
                return false;
            }
        }
        return true;
    }

    /**
     * Pretty-prints the Hamiltonian Cycle.
     */
    private static void printCycle(int[] path) {
        System.out.print("\n  Hamiltonian Cycle found:\n  ");
        for (int vertex : path) {
            System.out.print(vertex + " → ");
        }
        System.out.println(path[0] + "  (back to start)\n");
    }

    /**
     * Prints the adjacency matrix of the graph.
     */
    private static void printGraph(int[][] graph) {
        int n = graph.length;
        System.out.println("\n  Adjacency Matrix (" + n + " vertices):");
        System.out.print("     ");
        for (int i = 0; i < n; i++) System.out.printf("%2d ", i);
        System.out.println();
        System.out.print("     ");
        System.out.println("─".repeat(n * 3));
        for (int i = 0; i < n; i++) {
            System.out.printf("  %d │ ", i);
            for (int j = 0; j < n; j++) System.out.printf("%2d ", graph[i][j]);
            System.out.println();
        }
    }

    // ── MAIN ─────────────────────────────────────────────────
    public static void main(String[] args) {
        System.out.println("╔══════════════════════════════════╗");
        System.out.println("║        Hamiltonian Cycle         ║");
        System.out.println("╚══════════════════════════════════╝");

        // ── Graph 1: Hamiltonian Cycle EXISTS ────────────────
        // Cycle: 0 → 1 → 2 → 4 → 3 → 0
        int[][] graph1 = {
            {0, 1, 0, 1, 0},   // vertex 0 connects to 1, 3
            {1, 0, 1, 1, 1},   // vertex 1 connects to 0, 2, 3, 4
            {0, 1, 0, 0, 1},   // vertex 2 connects to 1, 4
            {1, 1, 0, 0, 1},   // vertex 3 connects to 0, 1, 4
            {0, 1, 1, 1, 0}    // vertex 4 connects to 1, 2, 3
        };

        System.out.println("\n─── Example 1: Cycle exists ───");
        printGraph(graph1);
        solveHamiltonianCycle(graph1);

        // ── Graph 2: No Hamiltonian Cycle ────────────────────
        int[][] graph2 = {
            {0, 1, 0, 1, 0},
            {1, 0, 1, 1, 1},
            {0, 1, 0, 0, 1},
            {1, 1, 0, 0, 0},   // removed edge 3-4
            {0, 1, 1, 0, 0}    // removed edge 4-3
        };

        System.out.println("\n─── Example 2: No cycle ───");
        printGraph(graph2);
        solveHamiltonianCycle(graph2);
    }
}

File: hamiltonian_cycle.py
"""
============================================================
 HAMILTONIAN CYCLE — Python
============================================================
 Given an undirected graph (adjacency matrix), determine
 whether a Hamiltonian Cycle exists — a path that visits
 every vertex exactly once and returns to the start.

 Approach : Backtracking
 Time     : O((N-1)!)  — NP-Complete problem
 Space    : O(N)
============================================================
"""
from typing import List, Optional

NOT_PLACED = -1


def solve_hamiltonian_cycle(graph: List[List[int]]) -> Optional[List[int]]:
    """
    Finds a Hamiltonian Cycle in the graph, if one exists.

    Parameters
    ----------
    graph : N×N adjacency matrix  (1=edge, 0=no edge)

    Returns
    -------
    List of vertices forming the cycle, or None if none exists.
    """
    n = len(graph)
    path = [NOT_PLACED] * n
    path[0] = 0   # always start at vertex 0

    if _backtrack(graph, path, pos=1, n=n):
        return path
    return None


def _backtrack(graph: list, path: list, pos: int, n: int) -> bool:
    """
    Core backtracking function.

    Parameters
    ----------
    graph : Adjacency matrix
    path  : Current path of placed vertices
    pos   : Index of the next slot to fill
    n     : Total number of vertices
    """

    # ── BASE CASE ────────────────────────────────────────────
    # All vertices placed — verify edge from last vertex back to start
    if pos == n:
        return graph[path[pos - 1]][path[0]] == 1

    # ── TRY EACH VERTEX AS NEXT IN PATH ──────────────────────
    for vertex in range(1, n):   # vertex 0 is always the start

        # ── PHASE 1: CHOOSE ──────────────────────────────────
        if _is_safe(graph, path, vertex, pos):

            path[pos] = vertex   # place vertex

            # ── PHASE 2: EXPLORE ─────────────────────────────
            if _backtrack(graph, path, pos + 1, n):
                return True      # complete cycle found!

            # ── PHASE 3: BACKTRACK ───────────────────────────
            path[pos] = NOT_PLACED   # un-place

    return False   # no vertex works here


def _is_safe(graph: list, path: list, vertex: int, pos: int) -> bool:
    """
    Returns True if 'vertex' can be placed at path[pos].

    Two conditions:
    1. There must be an edge from path[pos-1] to vertex.
    2. vertex must not already be in path.
    """
    # Condition 1: edge must exist
    if graph[path[pos - 1]][vertex] == 0:
        return False
    # Condition 2: vertex not already placed
    return vertex not in path[:pos]


def _print_result(path: Optional[List[int]]) -> None:
    if path is None:
        print("  No Hamiltonian Cycle exists.")
    else:
        cycle = " → ".join(str(v) for v in path) + f" → {path[0]}"
        print(f"\n  Hamiltonian Cycle found:\n  {cycle}\n")


def _print_graph(graph: List[List[int]]) -> None:
    n = len(graph)
    print(f"\n  Adjacency Matrix ({n} vertices):")
    header = "     " + "  ".join(str(i) for i in range(n))
    print(header)
    print("     " + "─" * (n * 3))
    for i, row in enumerate(graph):
        print(f"  {i}  │  " + "  ".join(str(v) for v in row))


# ── MAIN ─────────────────────────────────────────────────────
if __name__ == "__main__":
    print("╔══════════════════════════════════╗")
    print("║        Hamiltonian Cycle         ║")
    print("╚══════════════════════════════════╝")

    # Graph 1 — cycle exists: 0 → 1 → 2 → 4 → 3 → 0
    graph1 = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 1, 1],
        [0, 1, 0, 0, 1],
        [1, 1, 0, 0, 1],
        [0, 1, 1, 1, 0]
    ]
    print("\n─── Example 1: Cycle exists ───")
    _print_graph(graph1)
    result1 = solve_hamiltonian_cycle(graph1)
    _print_result(result1)

    # Graph 2 — no cycle (edge 3-4 removed)
    graph2 = [
        [0, 1, 0, 1, 0],
        [1, 0, 1, 1, 1],
        [0, 1, 0, 0, 1],
        [1, 1, 0, 0, 0],
        [0, 1, 1, 0, 0]
    ]
    print("─── Example 2: No cycle ───")
    _print_graph(graph2)
    result2 = solve_hamiltonian_cycle(graph2)
    _print_result(result2)
